### PR TITLE
[#6906] Added support for PostgreSQL ON CONFLICT .. ON CONSTRAINT ..

### PR DESCRIPTION
**Pull request by timur-sh**

### CODE REVIEW

**Comment by lukaseder** [comment]
Thanks for your suggestion. I'll review this soon.

File: jOOQ/src/main/java/org/jooq/impl/InsertImpl.java [hunk 1]
```diff
@@ -604,6 +609,27 @@ public final InsertImpl doNothing() {
...
+    @Override
+    public final InsertImpl onConflictOnConstraint(Name constraint) {
+        onConflictOnConstraint(constraint(constraint));
+        return this;
+    }
+
+    @Override
+    public final InsertImpl onConflictOnConstraint(UniqueKey<R> constraint) {
+        if (constraint.getName() == null)
+            throw new IllegalStateException("UniqueKey's name is not specified");
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertImpl.java]: Good idea. I'll switch this to `StringUtils.isEmpty()` (as empty names are possible too, and are also illegal here). Also, `IllegalArgumentException` seems more appropriate in this particular case

Comment (lukaseder) [reply jOOQ/src/main/java/org/jooq/impl/InsertImpl.java]: ... I've noticed just now that the exception is also thrown from `InsertQueryImpl` (which makes sense, of course). So, to remove repetition of code, I will change the above `InsertImpl` logic to simply delegate to `InsertQueryImpl` and let the latter handle the plumbing / data type conversion.

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java [hunk 2]
```diff
@@ -248,30 +273,41 @@ final void accept0(Context<?> ctx) {
...
-                       .start(INSERT_ON_DUPLICATE_KEY_UPDATE)
-                       .visit(K_ON_CONFLICT)
-                       .sql(" (");
+                       .start(INSERT_ON_DUPLICATE_KEY_UPDATE);
 
-                    if (onConflict != null && onConflict.size() > 0) {
-                        boolean qualify = ctx.qualify();
+                    if (onConstraint != null) {
+                        ctx.data(DATA_CONSTRAINT_REFERENCE, true);
+                        ctx.visit(K_ON_CONSTRAINT)
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: I suspect the `ON CONSTRAINT` keywords here should be prefixed by `ON CONFLICT`. The current implementation seems to produce a syntax error. I'll fix this when implementing the relevant integration tests.

**Comment by lukaseder** [comment]
Cool, no regressions in the integration tests... Now adding integration tests for the new functionality

File: jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java [hunk 3]
```diff
@@ -335,18 +371,30 @@ else if (onDuplicateKeyIgnore) {
...
                     ctx.formatSeparator()
-                       .start(INSERT_ON_DUPLICATE_KEY_UPDATE)
-                       .visit(K_ON_CONFLICT)
-                       .sql(' ');
+                       .start(INSERT_ON_DUPLICATE_KEY_UPDATE);
 
-                    if (onConflict != null && onConflict.size() > 0) {
-                        boolean qualify = ctx.qualify();
+                    if (onConstraint != null) {
+                        ctx.data(DATA_CONSTRAINT_REFERENCE, true);
```
Comment (lukaseder) [inline jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: Nice one. Curious about this, how did you discover this flag?

Comment (timur-sh) [reply jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: I've investigated `ConstraintImpl#accept` method. As I see, it's perfectly appropriate in use cases, when it's need to get only `name` of `Constraint`'s object by `Context`

Comment (lukaseder) [reply jOOQ/src/main/java/org/jooq/impl/InsertQueryImpl.java]: Yeah, it's an "acceptable hack" - a workaround if either the rendering API or the data model isn't sophisticated enough to communicate such a "special case" and the special case isn't worth an API change...

File: jOOQ/src/main/java/org/jooq/impl/InsertImpl.java (hunk 1 above)
Comment (timur-sh) [reply jOOQ/src/main/java/org/jooq/impl/InsertImpl.java]: I agree, using `StringUtils.isEmpty()` in this case is more appropriate then checking on `null`

**Pull request merged**